### PR TITLE
[Python]Sentry error reporting & Global config manager

### PR DESCRIPTION
**Pull request by AyushExel**
Conditions required to send errors (ALL conditions must be met or no errors will be reported):
        - sentry_sdk package is installed (Maybe we make it a dependency?)
        - sync=True in  settings
        - pytest is not running
        - running in a pip package installation
        - running in a non-git directory
        - online environment
 
- CLI config commands are needed to allow the user to be able to edit them, but maybe it'll be too large for this PR.

TODOs:
- [x] remove verbose sentry log msg (next PR)
- [x] needs lancedb.__version__ to tag the version of the error origin

### CODE REVIEW

File: python/lancedb/utils/config.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+        if not self.file.exists():
+            self.save()
+
+        self.load()
+        correct_keys = self.keys() == self.defaults.keys()
+        correct_types = all(
+            type(a) is type(b) for a, b in zip(self.values(), self.defaults.values())
+        )
+        if not (correct_keys and correct_types):
```
Comment (rok) [inline python/lancedb/utils/config.py:82]: Would this be an improvement or am I missing the point? :D
```suggestion
        config_created = False
        if not self.file.exists():
            self.save()
            config_created = True

        self.load()
        correct_keys = self.keys() == self.defaults.keys()
        correct_types = all(
            type(a) is type(b) for a, b in zip(self.values(), self.defaults.values())
        )
        if not (config_created and correct_keys and correct_types):
```

File: python/lancedb/utils/sentry_log.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            return
+
+        sentry_sdk.init(
+            dsn="https://2a892530e0c92d3a3194cddc897a9090@o4505873570791424.ingest.sentry.io/4505873580425216",
+            debug=False,
+            traces_sample_rate=1.0,
+            environment="production",  # 'dev' or 'production'
+            before_send=before_send,
+            ignore_errors=[KeyboardInterrupt, FileNotFoundError],
+        )
```
Comment (rok) [inline python/lancedb/utils/sentry_log.py:84]: I forget how tags work in sentry, but will be able to filter by lance/lancedb versions in sentry or should we add something here?

**Review by rok — COMMENTED** [review]
Great stuff! Added some small comments.
I used sentry in the past and it was great to quickly triage python errors!

File: python/lancedb/utils/config.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+import os
+import platform
+import uuid
+from pathlib import Path
+
+from .general import LOGGER, is_dir_writeable, yaml_load, yaml_save
+
+MACOS, LINUX, WINDOWS = (
+    platform.system() == x for x in ["Darwin", "Linux", "Windows"]
+)  # environment booleans
```
[the remark this answers is not in the document]
Comment (AyushExel) [reply python/lancedb/utils/config.py:12]: Yeah makes sense. I think they might be used for event tracking with posthog but for now eval on fly sounds better

File: python/lancedb/utils/config.py (hunk 1 above)
Comment (AyushExel) [reply python/lancedb/utils/config.py:82]: I see. I think if `config_created` is true, the keys and types will always be correct. This condition is aimed at re-creating the config file if the default configs are ever changed in the package. So, I think if config_created == True, then theres no need to check for types and keys, making it a redundant check? Is there some other condition you can think of?

File: python/lancedb/utils/sentry_log.py (hunk 2 above)
Comment (AyushExel) [reply python/lancedb/utils/sentry_log.py:84]: yeah version and other details are tracked here in `before_send` callback https://github.com/lancedb/lancedb/pull/492/files#diff-603e1842b90966a1a2eb9e41f61c62a6db0fd6b3d5a71cec2f0355387c3d7eb7R62

File: python/lancedb/utils/config.py (hunk 1 above)
Comment (rok) [reply python/lancedb/utils/config.py:82]: That would be redundant indeed. Ok nevermind.

**Review by wjones127 — CHANGES_REQUESTED** [review]
Noticed a few small issues.

Is there a link where we can see some examples of the telemetry data?

File: python/lancedb/utils/general.py [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+    "Colab"
+    if is_colab()
+    else "Kaggle"
+    if is_kaggle()
+    else "Jupyter"
+    if is_jupyter()
+    else "Docker"
+    if is_docker()
+    else platform.system()
+)
```
Comment (rok) [inline python/lancedb/utils/general.py:309]: If multiple of these are true (given the checks it wouldn't be too hard) we will only get the first one that was checked reported. Is the order of these checks ok or would we rather check all and concatenate into one string to capture all?

Comment (AyushExel) [reply python/lancedb/utils/general.py:309]: I think these should be mutually exclusive no? Like you cannot open colab inside of a kaggle notebook or open a kaggle kernel in a jupyter env?

Comment (rok) [reply python/lancedb/utils/general.py:309]: Sure, but results of checks might not be, for instance if you run jupyter server inside a docker container what would then be the most relevant environment to report to sentry?

Comment (AyushExel) [reply python/lancedb/utils/general.py:309]: I see.. So ENVIRONMENTS capturing all of them would be best then?
Something like this is okay or comma separated would be better?
```
ENVIRONMENTS = platform.system()
if is_colab():
    ENVIRONMENTS += " Colab "
if is_kaggle():
    ENVIRONMENTS += " Kaggle "
if is_jupyter():
    ENVIRONMENTS += " Jupyter "
if is_docker():
    ENVIRONMENTS += " Docker "
```

**Comment by AyushExel** [comment]
@wjones127 looks like this, I cannot invite anyone as it's not a paid account. I'll ask Chang to create a teams account.
Screenshot 2023-09-27 at 10 58 06 AM

**Comment by AyushExel** [comment]
@wjones127 sent an invite. You can see the issues here https://lancedb-00.sentry.io/issues/?referrer=sidebar

File: python/lancedb/utils/general.py [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+
+PLATFORMS = platform.system()
+if is_colab():
+    PLATFORMS += " Colab "
+if is_kaggle():
+    PLATFORMS += " Kaggle "
+if is_jupyter():
+    PLATFORMS += " Jupyter "
+if is_docker():
+    PLATFORMS += " Docker "
```
Comment (rok) [inline python/lancedb/utils/general.py:298]: How about:
```suggestion
PLATFORMS = [platform.system()]
if is_colab():
    PLATFORMS.append("Colab")
if is_kaggle():
    PLATFORMS.append("Kaggle")
if is_jupyter():
    PLATFORMS.append("Jupyter")
if is_docker():
    PLATFORMS.append("Docker")

PLATFORMS = "|".join(PLATFORMS)
```

Comment (rok) [reply python/lancedb/utils/general.py:298]: Sorry for  the bikeshedding :D

**Review by rok — APPROVED** [review]
LGTM!

**Comment by AyushExel** [comment]
Merging this in a different branch. where other telemetry PRs will be merged as well. Once ready, we can add everything to main

**Pull request merged**